codegenerator/cli/npm/envio/src/LazyLoader.res

```diff
@@ -0,0 +1,133 @@
+exception LoaderTimeout(string)
+
+type rec asyncMap<'key, 'value> = {
+  // The number of loaded results to keep cached. (For block loading this should be our maximum block interval)
+  _cacheSize: int,
+  // The maximum number of results we can try to load simultaneously.
+  _loaderPoolSize: int,
+  // How long to wait before retrying failures
+  // TODO: Add randomized exponential back-off (outdated)
+  _retryDelayMillis: int,
+  // How long to wait before cancelling a load request
+  _timeoutMillis: int,
+  // The promises we return to callers. We satisfy them asynchronously.
+  externalPromises: Utils.Map.t<'key, promise<'value>>,
+  // The handled used to populate the external promises once we have loaded their data.
+  resolvers: Utils.Map.t<'key, 'value => unit>,
+  // The keys currently being loaded
+  inProgress: Utils.Set.t<'key>,
+  // Keys  for items that we have not started loading yet.
+  loaderQueue: SDSL.Queue.t<'key>,
+  // Keys for items that have been loaded already. Used to evict the oldest keys from cache.
+  loadedKeys: SDSL.Queue.t<'key>,
+  // The function used to load the result.
+  loaderFn: 'key => promise<'value>,
```

Comment:
Having key as an unboxed variant would make the `loaderFn` annoying for usage, so I've decided to use `Map` to support literally any 'key type.

Reply:
Cool, that works 🙂

I still think just passing `toString` dependency is the best because it can avoid any confusion about passing reference vs a new object etc. And I'm sure Js.Dict is more performant than Map etc

Reply:
Makes sense. But I assume even thoug Js.Dict.t more performant than Map, but after adding toString there shouldn't be a difference.